### ISSUE TITLE
6 processes where a delimiter for text files can be chosen (f.e. }bedrock.cube.data.export)

### ISSUE THREAD

**Issue by wimgielis** [cubewise-code/bedrock#29]
**Describe the bug**
```
# Validate file delimiter & quote character
If( pDelimiter @= '' );
    sMessage = 'Data delimiter for file not specified.';
    nErrors = nErrors + 1;
    LogOutput( cMsgErrorLevel, Expand( cMsgErrorContent ) );
```

The Parameters section of the processes mentions that a comma is the default delimiter. Yet, as we can see in the code, if the supplied delimiter is empty, the process stops. The default comma is not applied. Either the text is wrong, either the code is wrong. I suspect that the code is wrong because the parameter question includes 'OPTIONAL' too, indicating that the default is a comma and the the processes should not abort.

It's probably down to personal preferences but I would like to replace this bit of code:

```
    # If length of pFieldDelim is exactly 3 chars and each of them is decimal digit, then the pFieldDelim is entered as ASCII code
    nValid = 0;
    If ( LONG(pFieldDelim) = cLenASCIICode );
      nChar = 1;
      While ( nChar <= cLenASCIICode );
        If( CODE( pFieldDelim, nChar ) >= CODE( '0', 1 ) & CODE( pFieldDelim, nChar ) <= CODE( '9', 1 ) );
          nValid = 1;
        Else;
          nValid = 0;
        EndIf;
        nChar = nChar + 1;
      End;
    EndIf;
    If ( nValid<>0 );
      pFieldDelim=CHAR(StringToNumber( pFieldDelim ));
    Else;
      pFieldDelim = SubSt( Trim( pFieldDelim ), 1, 1 );
    EndIf;
```

by:

```
    nValid = 0;
    If ( LONG(pFieldDelim) = cLenASCIICode );
      nChar = 1;
      While ( nChar <= cLenASCIICode );
        If( SCAN( SubSt( pFieldDelim, nChar, 1 ), '0123456789' ) > 0 );
          nValid = 1;
        Else;
          nValid = 0;
          Break;
        EndIf;
        nChar = nChar + 1;
      End;
    EndIf;
    If ( nValid=1 );
      pFieldDelim=CHAR(StringToNumber( pFieldDelim ));
    Else;
      pFieldDelim = SubSt( pFieldDelim, 1, 1 );
    EndIf;
```

This means 3 changes:
- using the Scan function towards '0123456789'
- the Break function stops a loop before the end. Every cycle counts :-)
- the Trim in pFieldDelim = SubSt( Trim( pFieldDelim ), 1, 1 ); is superfluous, given earlier code

**Comment by wimgielis** [comment]
Test to know if I can reply to conversations by email.
Post can be deleted afterwards.

------

Best regards / Beste groeten,

Wim Gielis
MS Excel MVP 2011-2014
http://www.wimgielis.com <http://www.wimgielis.be>


Op do 2 mei 2019 om 09:36 schreef Wim Gielis <notifications@github.com>:

**Comment by lotsaram** [comment]
@wimgielis 
If you would like to create a pull request to both fix the empty delimiter and improve the ascii numeric code to ascii character conversion then we would be happy to welcome you as a contributor.

**Comment by wimgielis** [comment]
I can do that for the TM1 changes. 2 questions:
- should , be a default or is the parameter description wrong ?
- I don't know the workings of github, so if there is a manual / description of pulling, merge, ... I would be happy to go through that material. I don't want to break things ;-)

**Comment by lotsaram** [comment]
Down the very bottom of the ReadMe file is 2 links describing how to fork a repository and the create a pull request back to the original one.

If changes are submitted by pull request then an admin of the repository always gets an opportunity to view the diff for each changed file and approve the changes before merging. So risk of breaking is very small via pull request.

In this case the documentation is correct and the error handling for empty delimiter is too strict.
In case of empty delimiter these should be the defaults:
* _string list delimiter_ & Ascii code 38
* _element start delimiter within filter string_ ¦ Ascii code 166 (delimiter between dimension name & 1st element)
* _element delimiter within filter string_ + Ascii code 43 (delimiter between element names of same dimension)
* _dimension delimiter within filter string_ & Ascii code 38 (delimiter between last element of previous dimension & next dimension name)
* _ascii field delimiter_ , Ascii code 44
* _ascii decimal delimiter_ , Ascii code 46

_Note:_ In any process with parameter inputs for _ascii thousand delimiter_ or _ascii quote character_ these should pass on blank character if that is what is passed as it's easy to think of use cases where you don't ant the output fields encapsulated in quotes or for number strings to contain thousand separators.

**Comment by lotsaram** [comment]
@mmastrolonardo can you take this one?

**Comment by mmastrolonardo** [comment]
The following following processes accepting optional delimiters have been reviewed and fixed:

}bedrock.cube.create
}bedrock.cube.data.copy
}bedrock.cube.data.export
}bedrock.cube.data.import
}bedrock.cube.rule.manage
}bedrock.cube.view.delete
}bedrock.cube.viewandsubsets.create
}bedrock.dim.attr.importfromfile
}bedrock.dim.clone
}bedrock.dim.delete
}bedrock.hier.create
}bedrock.hier.delete
}bedrock.hier.element.create
}bedrock.hier.export
}bedrock.hier.import
}bedrock.hier.sub.exporttofile

**Comment by lotsaram** [comment]
Hi @mmastrolonardo great! But fixed where? Can you please fork the master branch, then sync the changes to your fork, then create a pull request to the master?

**Comment by lotsaram** [comment]
addressed by #42
